Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -332,3 +391,132 @@ private func sandboxProfile() -> String {
     stream <<< ")" <<< "\n"
     return stream.bytes.asString!
 }
+
+// MARK:- Caching support.
+
+final class ManifestCacheDelegate: LLBuildEngineDelegate {
+
+    weak var loader: ManifestLoader!
+
+    func lookupRule(rule: String, key: Key) -> Rule {
+        switch rule {
+        case ManifestLoadRule.ruleName:
+            return ManifestLoadRule(key, loader: loader)
+        case FileInfoRule.ruleName:
+            return FileInfoRule(key)
+        case SwiftPMVersionRule.ruleName:
+            return SwiftPMVersionRule()
+        default:
+            fatalError("Unknown rule \(rule)")
+        }
+    }
+}
+
+/// A rule to load a package manifest.
+///
+/// The rule can currently only load manifests which are physically present on
+/// the local file system. The rule will re-run if the manifest is modified.
+final class ManifestLoadRule: LLBuildRule {
+
+    struct RuleKey: LLBuildKey {
+        typealias BuildValue = RuleValue
+        typealias BuildRule = ManifestLoadRule
+
+        let path: AbsolutePath
+        let manifestVersion: ManifestVersion
+    }
+
+    struct RuleValue: LLBuildValue {
+        let result: Result<String, StringError>
+    }
+
+    override class var ruleName: String { return "\(ManifestLoadRule.self)" }
+
+    private let key: RuleKey
+    private weak var loader: ManifestLoader!
+
+    init(_ key: Key, loader: ManifestLoader) {
+        self.key = RuleKey(key)
+        self.loader = loader
+        super.init()
+    }
+
+    override func start(_ engine: LLTaskBuildEngine) {
+        engine.taskNeedsInput(SwiftPMVersionRule.RuleKey(), inputID: 1)
+        engine.taskNeedsInput(FileInfoRule.RuleKey(path: key.path), inputID: 2)
```

Comment:
I think it'd be nice if the key was primary, so this looked like:

```
taskNeedsInput(FileInfoKey(path: path))
```

optimizing the clients. NBD though